magento/magento2
# Updated the error reporting for the Config Element Iterator.

### Description

Attempt to make the debugging a little easier when the system configuration is incorrect.
Needed to add the App state class and the Logger to determine how much logging is used and how to get the error to the end developer or user.
Updated the PHPUnit test to have logger and state mocks. Included a Unit test for confirming the exception message in developer mode.

### Fixed Issues (if relevant)

1. N/A

### Manual testing scenarios

0. Enable developer mode
1.  Pull in module with invalid system config
2. Try navigating to the Stores -> Configuration
3. Ensure there is a clear exception being thrown explaining which element key is breaking

### Contribution checklist
 - [ x ] Pull request has a meaningful description of its purpose
 - [ x ] All commits are accompanied by meaningful commit messages
 - [ x ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ? ] All automated tests passed successfully (all builds on Travis CI are green)
